library/std/src/net/ip.rs

```diff
@@ -1884,10 +1899,19 @@ impl From<[u16; 8]> for IpAddr {
 // Tests for this module
 #[cfg(all(test, not(target_os = "emscripten")))]
 mod tests {
+    use crate::mem::size_of;
     use crate::net::test::{sa4, sa6, tsa};
     use crate::net::*;
     use crate::str::FromStr;
 
+    #[test]
+    fn ipv6bytes_all_fields_have_the_same_size() {
+        use super::Ipv6Bytes;
+        let _addr = Ipv6Bytes { addr: [0_u8; 16] };
+        let _addr16 = Ipv6Bytes { addr16: [0_u16; 8] };
+        assert_eq!(size_of::<Ipv6Bytes>(), size_of::<[u8; 16]>());
```

Comment:
I wonder if `size_of_val` works here? Like `size_of_val(&addr)` and `size_of_val(&addr16)`.

Reply:
Does they have the same type?

Reply:
Yes, they have the same type. If they have the same type it means they have the same size? I am not sure the behavior of `size_of_val` for a `union`. Looks like we could do that with unsafe. Let me send a suggestion.

Prototype: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8c5f14837cefa7cc61951bd338278eec

Reply:
```suggestion
        unsafe {
            assert_eq!(size_of_val(&addr.addr), size_of_val(&addr16.addr16));
        }
```

Reply:
> If they have the same type it means they have the same size? 
Yes, if not that's would be the most unsound issue.

The two `let` lines above will fail at complie time if the field changes its size.

Reply:
> Yes, if not that's would be the most unsound issue.

That is what we want to check in the test here rather than typing it out as text in comment since the type system does not seemed to be able to validate this.

> The two let lines above will fail at complie time if the field changes its size.

Yes, that is what we want to check, both the fields should have the same size.

Reply:
Comparing `size_of_val` on two instances of the union will always be equal, but comparing on each field makes sense.

Reply:
> Comparing size_of_val on two instances of the union will always be equal, but comparing on each field makes sense.

@cuviper Yes, I did a test in nightly to check, comparing the same type give the same output but however when I compare the fields, it gives different output. We could use that here, I rather it does not need unsafe though but seemed like we need it.